src/voter/mod.rs

```diff
@@ -888,17 +884,17 @@ mod tests {
 					last_round_state,
 					last_finalized,
 				);
-				::tokio::spawn(exit.clone()
-					.until(voter.map_err(|_| panic!("Error voting"))).map(|_| ()));
+				threads_pool.spawn_ok(voter.map(|v| v.expect("Error voting")));
 
 				// wait for the best block to be finalized by all honest voters
 				finalized
-					.take_while(|&(_, n, _)| Ok(n < 6))
-					.for_each(|_| Ok(()))
-			});
+					.take_while(|&(_, n, _)| future::ready(n < 6))
+					.for_each(|_| future::ready(()))
+			}).collect::<Vec<_>>();
 
-			::futures::future::join_all(finalized_streams).map(|_| signal.fire())
-		})).unwrap();
+			threads_pool.spawn_ok(routing_task.map(|_| ()));
```

Comment:
Note: Because the behaviour of `join_all` changed, I had to adjust the order of initialization, otherwise the test wouldn't work (blocked at round 1 with futures not being notified). This is a bit concerning, but I couldn't find the origin of the issue.

Reply:
From what I can tell this is not related to `join_all`. If the `routing_task` future is spawned before the first `Voter` is created (i.e. the case where the test stalls / hangs), the `NetworkRouting` future is polled "immediately" before the first voter is created. It then returns `Poll::Pending`, which it always does, since it relies on task wakeup from the `UnboundedReceiver`s (i.e. when messages are sent). However, there are no voters and hence no `VotingRound` yet, so the current task is only scheduled to  be polled again if a message is sent to the `GlobalMessageNetwork`, which doesn't happen in this test. So the `NetworkRouting` is never polled again, even after the first round with voters is created, resulting in the messages never being dispatched.

In all other tests except one (`skips_to_latest_round_after_catch_up`), the `routing_task` is only spawned after the first voter is created. In `skips_to_latest_round_after_catch_up`, the it is spawned earlier, but this test then calls `network.send_message` after a voter is set up, i.e. sends a message on the `GlobalMessageNetwork`, resulting in the `NetworkRouting` to be polled again.

The reason this worked previously is most likely due to the semantics of `block_on_all` in tokio, which results in the spawned "inner" `NetworkRouting` future to be polled again after the "outer" future resolved, at which point the voters have been set up and a voting round has been created.

So it seems to me that spawning the `routing_task` only after at least one voter has been created is not only necessary now but arguably more correct, at least the way the `NetworkRouting` is currently set up.

Reply:
Thanks for looking into this and for the detailed explanation @romanb.